### ISSUE TITLE
Can't use labels argument in `scale_*()` when `show.limits = TRUE`

### ISSUE THREAD

**Issue by brunomioto** [tidyverse/ggplot2#5009]
It always gives 
```
Error in `check_breaks_labels()`:
! `breaks` and `labels` must have the same length
```

Reprex
``` r
library(tidyverse)

#works
mtcars %>%
  ggplot(aes(hp, mpg, color = mpg)) +
  geom_point() +
  scale_color_stepsn(
    colors = c("#dd3497",
               "#ae017e",
               "#7a0177",
               "#49006a"),
    breaks = c(15,20,25),
    #labels = c("<15","15 text","20 text","25 text",">25")
  )+
  guides(color = guide_colorsteps(show.limits = TRUE))
```

![](https://i.imgur.com/vExNIIg.png)

``` r
#doesn't work
mtcars %>%
  ggplot(aes(hp, mpg, color = mpg)) +
  geom_point() +
  scale_color_stepsn(
    colors = c("#dd3497",
               "#ae017e",
               "#7a0177",
               "#49006a"),
    breaks = c(15,20,25),
    labels = c("<15","15 text","20 text","25 text",">25")
  )+
  guides(color = guide_colorsteps(show.limits = TRUE))
#> Error in `check_breaks_labels()`:
#> ! `breaks` and `labels` must have the same length
```

<sup>Created on 2022-10-08 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

![](https://i.imgur.com/HB6udFM.png)

**Comment by teunbrand** [comment]
I agree that this is a bit finnicky, but one way you could get the results you're after is to manually set limits, and include those limits in the breaks.

``` r
library(ggplot2)

ggplot(mtcars, aes(hp, mpg, color = mpg)) +
  geom_point() +
  scale_color_stepsn(
    colors = c("#dd3497",
               "#ae017e",
               "#7a0177",
               "#49006a"),
    limits = c(5, 35),
    breaks = c(5, 15,20,25, 35),
    labels = c("<15","15 text","20 text","25 text",">25")
  )
```

![](https://i.imgur.com/HpgtWLM.png)

<sup>Created on 2022-12-15 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by r2evans** [comment]
@teunbrand, there have been several questions related to `scale_*_stepsn` where the addition of `limits=` seems to resolve it. Further, the solutions I've tested all seem to work fine with `limits=range(..breaks..) + c(-1,1)` (pseudocode). It would seem logical to me that this may be the intuitive behavior. (Most recent: https://stackoverflow.com/q/77807880.)

I've tried tracing into `binned_scale` and its `ggproto` and don't have the proficiency atm to determine where `limits=` are in-play. Are you able to explain how `limits=NULL` is defined internally in the context of `scale_*_stepsn(.., colors=, breaks=)`? If not (or if it is relatively "arbitrary"), is there the ability to adjust the default behavior so that `range(..breaks..)+c(-1,1)` is inferred?

**Comment by teunbrand** [comment]
How the limits of a binned scale are defined is a bit awkward, as they're updated once the breaks are calculated.
The way I understand them is as follows:

If `limits = NULL`, a temporary 'candidate' limits equal to the data range is used for calculating the breaks.
Then, if original `limits = NULL`,  new limits are instated based on the breaks. The new lower limit is equal to `breaks[1] + (breaks[1] - breaks[2])` or in words, 'the first break minus the first step size' and the new upper limit is equal `breaks[n] + (breaks[n] - breaks[n - 1])` or 'the last breaks plus the last step size'. That happens here (where `limits` is the temporary one and `self$limits` is user input):

https://github.com/tidyverse/ggplot2/blob/a4be39dee07010c67c31ace059b1f0a24129fc65/R/scale-.R#L1202-L1230

**Comment by teunbrand** [comment]
The second example is because your first bin gets the midpoint of `(-80 + 10) / 2 == -35` and the last bin gets the midpoint `(190 + 100) / 2 == 145`. These will get the colour value of `scales::rescale(c(-35, 145), from = c(-80, 190))` ~= 16% and 83% along the full gradient.

Relevant for the first example, and what I forgot to mention, is that the limit updating thing only kicks in if there were no manually provided breaks. You can debug ggproto methods like so:

```r
debugonce(environment(ScaleBinned$get_breaks)$f)
```

**Comment by teunbrand** [comment]
I'm going to go ahead and say that including the limits in the break is the solution to the issue.
It is already the suggested solution in the warning message when you have atomic labels combined with `show.limits = TRUE`.

``` r
library(ggplot2)
ggplot(mtcars, aes(hp, mpg, color = mpg)) +
  geom_point() +
  scale_color_stepsn(
    colors = c("#dd3497",
               "#ae017e",
               "#7a0177",
               "#49006a"),
    breaks = c(15,20,25),
    labels = c("15 text","20 text","25 text")
  )+
  guides(color = guide_colorsteps(show.limits = TRUE))
#> Warning: `show.limits` is ignored when `labels` are given as a character vector.
#> ℹ Either add the limits to `breaks` or provide a function for `labels`.
```

![](https://i.imgur.com/rAbI9bS.png)

<sup>Created on 2025-01-10 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>
